src/Functions/FunctionFile.cpp

```diff
@@ -0,0 +1,134 @@
+        const ColumnString * expected = checkAndGetColumn<ColumnString>(column.get());
+        if (!expected)
+            throw Exception(
+                fmt::format("Illegal column {} of argument of function {}", arguments[0].column->getName(), getName()),
+                ErrorCodes::ILLEGAL_COLUMN);
+
+        const ColumnString::Chars & chars = expected->getChars();
+        const ColumnString::Offsets & offsets = expected->getOffsets();
+
+        std::vector<String> checked_filenames(input_rows_count);
+
+        auto result = ColumnString::create();
+        auto & res_chars = result->getChars();
+        auto & res_offsets = result->getOffsets();
+
+        res_offsets.resize(input_rows_count);
+
+        size_t source_offset = 0;
+        size_t result_offset = 0;
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            const char * filename = reinterpret_cast<const char *>(&chars[source_offset]);
+
+            const String user_files_path = context.getUserFilesPath();
+            String user_files_absolute_path = Poco::Path(user_files_path).makeAbsolute().makeDirectory().toString();
+            Poco::Path poco_filepath = Poco::Path(filename);
+            if (poco_filepath.isRelative())
+                poco_filepath = Poco::Path(user_files_absolute_path, poco_filepath);
+            const String file_absolute_path = poco_filepath.absolute().toString();
+            checkReadIsAllowedOrThrow(user_files_absolute_path, file_absolute_path);
+
+            checked_filenames[row] = file_absolute_path;
+            auto file = Poco::File(file_absolute_path);
+
+            if (!file.exists())
+                throw Exception(fmt::format("File {} doesn't exist.", file_absolute_path), ErrorCodes::FILE_DOESNT_EXIST);
+
+            const auto current_file_size = Poco::File(file_absolute_path).getSize();
+
+            result_offset += current_file_size + 1;
+            res_offsets[row] = result_offset;
+            source_offset = offsets[row];
+        }
+
+        res_chars.resize(result_offset);
+
+        size_t prev_offset = 0;
+
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            auto file_absolute_path = checked_filenames[row];
+            ReadBufferFromFile in(file_absolute_path);
+            char * res_buf = reinterpret_cast<char *>(&res_chars[prev_offset]);
+
+            const size_t file_lenght = res_offsets[row] - prev_offset - 1;
+            prev_offset = res_offsets[row];
+            in.readStrict(res_buf, file_lenght);
+            res_buf[file_lenght] = '\0';
+        }
+
+        return result;
+    }
+
+private:
+
+    void checkReadIsAllowedOrThrow(const std::string & user_files_absolute_path, const std::string & file_absolute_path) const
+    {
+        // If run in Local mode, no need for path checking.
+        if (context.getApplicationType() != Context::ApplicationType::LOCAL)
+            if (file_absolute_path.find(user_files_absolute_path) != 0)
```

Comment:
Does the `user_files_absolute_path` contain slash at the end?
Is it guaranteed?

Why do we use string processing functions instead of std::filesystem?
Do we have similar code in StorageFile?

Reply:
> Is is guaranteed? 

[I think yes](https://github.com/ClickHouse-Extras/poco/blob/fbaaba4a02e29987b8c584747a496c79528f125f/Foundation/src/Path_UNIX.cpp#L45)

> Why do we use string processing functions instead of std::filesystem?

I don't know how to check that directory is inside another directory with `std::filesystem` in a convenient way.  Maybe we can recursively call `parent_path` and check.

>Do we have similar code in StorageFile?
[Yes.]( https://github.com/nikitamikhaylov/ClickHouse/blob/84c7589c9a30a5536dc020345fcc5651ce517371/dbms/src/Storages/StorageFile.cpp#L112)